File: komorebi_desktop/src-tauri/src/commands/rag.rs
//! RAG (retrieval-augmented generation) folder management commands.

use std::sync::Arc;
use tauri::{AppHandle, Manager, State, Wry};

#[tauri::command]
pub fn rag_list_folders(
    rag: State<'_, Arc<komorebi_storage::RagIndex>>,
) -> Result<Vec<komorebi_storage::FolderStats>, String> {
    rag.folders().map_err(|e| e.to_string())
}

#[tauri::command]
pub fn rag_add_folder(
    rag: State<'_, Arc<komorebi_storage::RagIndex>>,
    path: String,
) -> Result<(), String> {
    rag.add_folder(std::path::Path::new(&path))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn rag_remove_folder(
    rag: State<'_, Arc<komorebi_storage::RagIndex>>,
    path: String,
) -> Result<(), String> {
    rag.remove_folder(std::path::Path::new(&path))
        .map_err(|e| e.to_string())
}

/// Re-walks either a single folder (if `path` is given) or every known
/// folder. Runs on a blocking thread so the UI stays responsive.
#[tauri::command]
pub async fn rag_reindex(
    app: AppHandle<Wry>,
    path: Option<String>,
) -> Result<komorebi_storage::IndexReport, String> {
    let rag: Arc<komorebi_storage::RagIndex> = app
        .try_state::<Arc<komorebi_storage::RagIndex>>()
        .ok_or_else(|| "RAG index not initialized".to_string())?
        .inner()
        .clone();
    tokio::task::spawn_blocking(move || {
        if let Some(p) = path {
            rag.index_folder(std::path::Path::new(&p))
        } else {
            rag.index_all()
        }
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())
}

File: dix_desktop/src/api/rag.ts
import { invoke } from "@tauri-apps/api/core";
import type { FolderStats, IndexReport } from "./types";

export async function setRagEnabled(enabled: boolean): Promise<void> {
  await invoke("set_rag_enabled", { enabled });
}

export async function ragListFolders(): Promise<FolderStats[]> {
  return invoke<FolderStats[]>("rag_list_folders");
}

export async function ragAddFolder(path: string): Promise<void> {
  await invoke("rag_add_folder", { path });
}

export async function ragRemoveFolder(path: string): Promise<void> {
  await invoke("rag_remove_folder", { path });
}

export async function ragReindex(path?: string): Promise<IndexReport> {
  return invoke<IndexReport>("rag_reindex", { path: path ?? null });
}
